File: main/java/com/example/wofertas/OfertaAdapter.java
package com.example.wofertas;

import android.content.Context;
import android.content.Intent;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Filter;
import android.widget.Filterable;
import android.widget.ImageView;
import android.widget.TextView;
import android.widget.Toast;
import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;
import com.bumptech.glide.Glide;
import java.util.ArrayList;
import java.util.List;
import java.util.Locale;

public class OfertaAdapter extends RecyclerView.Adapter<OfertaAdapter.OfertaViewHolder> implements Filterable {

    private List<Oferta> listaOfertasVisiveis;
    private List<Oferta> listaOfertasCompleta;
    private final Context context;

    public OfertaAdapter(List<Oferta> ofertas, Context context) {
        this.listaOfertasVisiveis = ofertas;
        this.listaOfertasCompleta = new ArrayList<>(ofertas);
        this.context = context;
    }

    public void atualizarListaCompleta(List<Oferta> novaLista) {
        this.listaOfertasCompleta = new ArrayList<>(novaLista);
        // Aplica o filtro atual à nova lista (se houver um) ou simplesmente reseta
        getFilter().filter(null);
    }

    @NonNull
    @Override
    public OfertaViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        View view = LayoutInflater.from(context).inflate(R.layout.item_oferta, parent, false);
        return new OfertaViewHolder(view);
    }

    @Override
    public void onBindViewHolder(@NonNull OfertaViewHolder holder, int position) {
        Oferta oferta = listaOfertasVisiveis.get(position);

        holder.txtNome.setText(oferta.getNome());
        holder.txtDescricao.setText(oferta.getDescricao());

        if (oferta.getDistancia() > 0 && oferta.getDistancia() != Double.MAX_VALUE) {
            holder.txtDistancia.setText(String.format(Locale.getDefault(), "%.1f km", oferta.getDistancia() / 1000.0));
            holder.txtDistancia.setVisibility(View.VISIBLE);
        } else {
            holder.txtDistancia.setVisibility(View.GONE);
        }

        if (oferta.getThumbUrl() != null && !oferta.getThumbUrl().isEmpty()) {
            Glide.with(context)
                    .load(oferta.getThumbUrl())
                    .placeholder(R.drawable.ic_placeholder)
                    .into(holder.imgOferta);
        } else {
            holder.imgOferta.setImageResource(R.drawable.ic_placeholder);
        }

        holder.itemView.setOnClickListener(v -> {
            String pdfUrl = oferta.getUrlPdf();
            if (pdfUrl != null && !pdfUrl.isEmpty()) {
                Intent intent = new Intent(context, VerPDF.class);
                intent.putExtra("oferta_pdf_url", pdfUrl);
                intent.putExtra("oferta_nome", oferta.getNome() + " - " + oferta.getDescricao());
                context.startActivity(intent);
            } else {
                Toast.makeText(context, "PDF indisponível", Toast.LENGTH_SHORT).show();
            }
        });
    }

    @Override
    public int getItemCount() {
        return listaOfertasVisiveis.size();
    }

    @Override
    public Filter getFilter() {
        return new Filter() {
            @Override
            protected FilterResults performFiltering(CharSequence constraint) {
                List<Oferta> filtradas = new ArrayList<>();
                if (constraint == null || constraint.length() == 0) {
                    filtradas.addAll(listaOfertasCompleta);
                } else {
                    String filtroPadrao = constraint.toString().toLowerCase().trim();
                    for (Oferta o : listaOfertasCompleta) {
                        boolean nomeContem = o.getNome() != null && o.getNome().toLowerCase().contains(filtroPadrao);
                        boolean descricaoContem = o.getDescricao() != null && o.getDescricao().toLowerCase().contains(filtroPadrao);
                        if (nomeContem || descricaoContem) {
                            filtradas.add(o);
                        }
                    }
                }
                FilterResults results = new FilterResults();
                results.values = filtradas;
                return results;
            }

            @Override
            protected void publishResults(CharSequence constraint, FilterResults results) {
                listaOfertasVisiveis.clear();
                if (results.values instanceof List) {
                    listaOfertasVisiveis.addAll((List<Oferta>) results.values);
                }
                notifyDataSetChanged();
            }
        };
    }

    static class OfertaViewHolder extends RecyclerView.ViewHolder {
        ImageView imgOferta;
        TextView txtNome, txtDescricao, txtDistancia;

        public OfertaViewHolder(@NonNull View itemView) {
            super(itemView);
            imgOferta = itemView.findViewById(R.id.imgOferta);
            txtNome = itemView.findViewById(R.id.txtNome);
            txtDescricao = itemView.findViewById(R.id.txtDescricao);
            txtDistancia = itemView.findViewById(R.id.txtDistancia);
        }
    }
}

File: app/src/main/java/com/example/wofertas/OfertaAdapter.kt
package com.example.wofertas

import android.annotation.SuppressLint
import android.content.Context
import android.location.Location
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Filter
import android.widget.Filterable
import android.widget.ImageView
import android.widget.TextView
import androidx.recyclerview.widget.RecyclerView
import com.bumptech.glide.Glide
import com.example.wofertas.utils.loadImage
import java.text.DecimalFormat
import java.util.Collections
import java.util.Locale

class OfertaAdapter(
    ofertasIniciais: List<Oferta>,
    private val listener: OnOfertaClickListener?
) : RecyclerView.Adapter<OfertaAdapter.OfertaViewHolder>(), Filterable {

    private val listaVisiveis: MutableList<Oferta> = ArrayList(ofertasIniciais)
    private val listaCompleta: MutableList<Oferta> = ArrayList(ofertasIniciais)
    private val context: Context? = listener as? Context
    private var ultimaLocalizacao: Location? = null

    interface OnOfertaClickListener {
        fun onOfertaClick(oferta: Oferta)
        fun onSaveClick(oferta: Oferta, position: Int)
    }

    fun setUltimaLocalizacaoUsuario(location: Location?) {
        this.ultimaLocalizacao = location
    }
    fun updateList(novaLista: List<Oferta>) {
        // 1. Atualiza a lista que serve de base para a busca
        this.listaCompleta.clear()
        this.listaCompleta.addAll(novaLista)

        // 2. Atualiza a lista que está sendo exibida na tela agora
        this.listaVisiveis.clear()
        this.listaVisiveis.addAll(novaLista)

        // 3. Avisa o Android para redesenhar a lista
        notifyDataSetChanged()
    }

    @SuppressLint("NotifyDataSetChanged")
    fun atualizarListaCompleta(novaLista: List<Oferta>) {
        listaCompleta.clear()
        listaCompleta.addAll(novaLista)
        // CORREÇÃO: Garante que a lista visível também receba os dados antes do filtro
        listaVisiveis.clear()
        listaVisiveis.addAll(novaLista)
        notifyDataSetChanged()
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): OfertaViewHolder {
        val view = LayoutInflater.from(parent.context).inflate(R.layout.item_oferta, parent, false)
        return OfertaViewHolder(view)
    }
    override fun onBindViewHolder(holder: OfertaViewHolder, position: Int) {
        val oferta = listaVisiveis[position]

        // 1. Dados de Texto
        holder.txtNome.text      = oferta.nome ?: "Oferta"
        holder.txtDescricao.text = "🏬 ${oferta.nomeSupermercado ?: "Supermercado"}"
        holder.txtDetalhesOferta.visibility = View.GONE

        // 2. Distância (Calculada via Localização)
        if (oferta.distancia < Double.MAX_VALUE && oferta.distancia > 0.0) {
            val df = DecimalFormat("#.#")
            holder.txtDistancia.text = "${df.format(oferta.distancia)} km"
            holder.txtDistancia.visibility = View.VISIBLE
        } else {
            holder.txtDistancia.visibility = View.GONE
        }

        // 3. Imagem (MUDANÇA: Use imagemOferta para mostrar a promoção em vez da logo)
        holder.imgOferta.loadImage(oferta.imagemOferta, R.drawable.logo_supermercado_placeholder)

        // 4. Coração (Favoritos - use isSaved do seu modelo)
        holder.imageViewSave.setImageResource(
            if (oferta.isSaved) R.drawable.ic_heart_filled else R.drawable.ic_heart_empty
        )

        // 5. Cliques
        holder.imageViewSave.setOnClickListener {
            // Envia o clique para a ListaOfertas tratar o salvamento no MongoDB
            val currentPosition = holder.adapterPosition
            if (currentPosition != RecyclerView.NO_POSITION) {
                listener?.onSaveClick(oferta, currentPosition)
            }
        }

        holder.itemView.setOnClickListener {
            // Envia o clique para a ListaOfertas usar o DataHolder e abrir o VerPDF
            listener?.onOfertaClick(oferta)
        }
    }


    override fun getItemCount(): Int = listaVisiveis.size

    override fun getFilter(): Filter = object : Filter() {
        override fun performFiltering(constraint: CharSequence): FilterResults {
            val filtradas: MutableList<Oferta> = ArrayList()
            val query = constraint.toString().lowercase(Locale.getDefault()).trim()

            if (query.isEmpty()) {
                filtradas.addAll(listaCompleta)
            } else {
                for (o in listaCompleta) {
                    val matchSuper = o.nomeSupermercado?.lowercase()?.contains(query) == true
                    val matchNome  = o.nome?.lowercase()?.contains(query) == true
                    if (matchSuper || matchNome) filtradas.add(o)
                }
            }
            return FilterResults().apply { values = filtradas }
        }

        @SuppressLint("NotifyDataSetChanged")
        override fun publishResults(constraint: CharSequence, results: FilterResults) {
            val temp = (results.values as? List<*> ?: emptyList<Any>())
                .filterIsInstance<Oferta>()
                .toMutableList()

            // ── 1. Cálculo de Distância ───────────────────────────────────────
            val locUsuario = ultimaLocalizacao
            temp.forEach { oferta ->
                oferta.distancia = if (locUsuario != null && oferta.latitude != null && oferta.longitude != null) {
                    val locMercado = Location("").apply {
                        latitude = oferta.latitude!!
                        longitude = oferta.longitude!!
                    }
                    locUsuario.distanceTo(locMercado) / 1000.0
                } else {
                    Double.MAX_VALUE
                }
            }

            // ── 2. Ordenação ──────────────────────────────────────────────────
            temp.sortWith(compareByDescending<Oferta> { it.isSaved }.thenBy { it.distancia })

            listaVisiveis.clear()
            listaVisiveis.addAll(temp)
            notifyDataSetChanged()
        }
    }

    class OfertaViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
        val imgOferta: ImageView        = itemView.findViewById(R.id.imgOferta)
        val txtNome: TextView           = itemView.findViewById(R.id.txtNome)
        val txtDescricao: TextView      = itemView.findViewById(R.id.txtDescricao)
        val txtDistancia: TextView      = itemView.findViewById(R.id.txtDistancia)
        val txtDetalhesOferta: TextView = itemView.findViewById(R.id.txtDetalhesOferta)
        val imageViewSave: ImageView    = itemView.findViewById(R.id.imageViewSave)
    }
}
